<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [47]:
import pandas as pd
import psycopg2

In [48]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [49]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [50]:
# текст запроса
query_3_1 = f'''select count(*)
                from public.vacancies
            '''

In [51]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [52]:
# текст запроса
query_3_2 = f'''select count(*)
                from public.employers
            '''

In [53]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [54]:
# текст запроса
query_3_3 = f'''select count(*)
                from public.areas
            '''

In [55]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [56]:
# текст запроса
query_3_4 = f'''select count(*)
                from public.industries
            '''

In [57]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# выводы по предварительному анализу данных
В данном юните была получена общая информация о данных - количество вакансий в базе данных, работодателей, регионов и сфер деятельности. Мы видим, что данные достаточно обширны и охватывают большое количество регионов (1362), достаточно разнообразное количество сфер деятельности (294). Работодателей заметно меньше, чем вакансий - это говорит о том, что большинство компаний выставляют несколько вакансий сразу (в среднем получается около 2 вакансий на работодателя).

Количество вакансий в базе - почти 50 тысяч (49197), количество работодателей - 23501. Такие обширные данные позволят нам провести качестванное исследование и обнаружить различные зависимости в дальнейшем.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [58]:
# текст запроса
query_4_1 = f'''select
                    a.name area,
                    count(*) cnt
                from public.areas a
                join public.vacancies v on a.id = v.area_id
                group by 1
                order by 2 DESC
                limit 15
            '''

In [59]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


area   cnt
0            Москва  5333
1   Санкт-Петербург  2851
2             Минск  2112
3       Новосибирск  2006
4            Алматы  1892
5      Екатеринбург  1698
6   Нижний Новгород  1670
7            Казань  1415
8         Краснодар  1301
9            Самара  1144
10   Ростов-на-Дону  1131
11          Воронеж  1063
12          Ташкент   851
13       Красноярск   847
14       Нур-Султан   840

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [60]:
# текст запроса
query_4_2 = f'''select
                    count(*)
                from public.vacancies
                where (salary_from is not null) or (salary_to is not null)
            '''

In [61]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [62]:
# текст запроса
query_4_3 = f'''select
                    round(avg(salary_from)) as from,
                    round(avg(salary_to)) as to
                from public.vacancies
            '''

In [63]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


from        to
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [64]:
# текст запроса
query_4_4 = f'''select
                    schedule,
                    employment,
                    count(*)
                from public.vacancies
                group by 1, 2
                order by 3 DESC
            '''

In [65]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [66]:
# текст запроса
query_4_5 = f'''select
                    experience,
                    count(*)
                from public.vacancies
                group by 1
                order by 2
            '''

In [67]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# выводы по детальному анализу вакансий
В первом задании было проанализированно количество вакансий в различных городах. Ожидаемо оказалось, что больше всего вакансий размещено в городах с самым большим населением (Москва, Санкт-Петербург). В других лидерах списка - города миллионники России и других стран (Беларуси, Казахстана, Узбекистана). Около половины всех вакансий из базы данных приходится на 15 крупнейших городов.

Далее были детально рассмотрены предлагаемые зарплаты. Мы увидели, что около половины вакансий выставлены без каких-либо указаний по зарплатам. Из тех вакансий, где зарплаты были указаны, средний диапазон составлял от 71 до 110 тыс.р.

Далее рассматривается популярность различных типов графиков и видов трудоустройства. Из анализа видно, что самым востребованным типом занятости является полная занятость - на нее приходится более 90% всех вакансий. Среди трех популярных графиков работы - полный день, удаленная работа и гибкий график, причем полный день лидирует с большим отрывом от остальных. Проектные работы, стажировки, вахтовый метод и сменный график - среди наименее популярных типов графиков и трудоустройств.

Анализ популярности опыта работы показывает, что более половины вакансий приходится на опыт от 1 до 3 лет, следующий за ним опыт - от 3 до 6 лет. Около 15% вакансий предлагают работу соискателям без опыта работы, и совсем небольшое число вакансий (1337, что составляет около 3% от общего числа вакансий) требует опыт работы более 6 лет.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [68]:
# текст запроса
query_5_1 = f'''select
                    e.name,
                    count(v.id)
                from public.vacancies v
                join public.employers e on v.employer_id = e.id
                group by 1
                order by 2 DESC
                limit 5
            '''

In [69]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [70]:
# текст запроса
query_5_2 = f'''select
                    a.name,
                    count(e.id) as count_employer,
                    count(v.id) as count_vacancies
                from public.areas a
                left join public.vacancies v  on v.area_id = a.id
                left join public.employers e on a.id  = e.area
                group by 1
                having count(v.id) = 0
                order by 2 DESC
                limit 1
            '''


In [71]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


name  count_employer  count_vacancies
0  Россия             410                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [72]:
# текст запроса
query_5_3 = f'''select
                    e.name,
                    count(distinct v.area_id)
                from public.vacancies v
                join public.employers e on v.employer_id = e.id
                group by 1
                order by 2 DESC
            '''

In [73]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
14761                  UniSol      1
14762            UNISTORY LLC      1
14763                   UNIT6      1
14764     United Distribution      1
14765              UNITEL LLC      1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [74]:
# текст запроса
query_5_4 = f'''select
                count(e.id)
                from public.employers e
                left join public.employers_industries ei on ei.employer_id = e.id
                where industry_id is null
            '''

In [75]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [76]:
# текст запроса
query_5_5 = f'''select
                    e.name,
                    count(industry_id)
                from public.employers e
                join public.employers_industries ei on ei.employer_id = e.id
                group by 1
                having count(industry_id) = 4
                order by 1
                limit 3
            '''

In [77]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


name  count
0  101 Интернет      4
1      21vek.by      4
2          2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [78]:
# текст запроса
query_5_6 = f'''select
                    count(ei.employer_id)
                from public.employers_industries ei
                join public.industries i on ei.industry_id = i.id
                where i.name = 'Разработка программного обеспечения'
            '''

In [79]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [108]:
# код для получения списка городов-милионников
import requests # Импортируем библиотеку requests
from bs4 import BeautifulSoup # Импортируем библиотеку BeautifulSoup
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8' # Определяем адрес страницы
response = requests.get(url) # Выполняем GET-запрос, содержимое ответа присваивается переменной response
page = BeautifulSoup(response.text, 'html.parser') # Создаём объект BeautifulSoup, указывая html-парсер
links = page.find_all('a')
list_of_cities = tuple([link.text for link in links[38:54]])
print(list_of_cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [109]:
# текст запроса
query_5_7 = f'''(select
            a.name,
            count(v.id)
            from public.vacancies v
            join public.employers e on v.employer_id = e.id
            join public.areas a on v.area_id = a.id
            where e.name = 'Яндекс'
            and a.name in {list_of_cities}
            group by 1)
            union
            (select
            'Total',
            count(v.id)
            from public.vacancies v
            join public.employers e on v.employer_id = e.id
            join public.areas a on v.area_id = a.id
            where e.name = 'Яндекс'
            and a.name in {list_of_cities})
            order by 2
            '''

In [110]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name  count
0              Омск     21
1         Челябинск     22
2        Красноярск     23
3         Волгоград     24
4             Пермь     25
5            Казань     25
6    Ростов-на-Дону     25
7               Уфа     26
8            Самара     26
9         Краснодар     30
10          Воронеж     32
11      Новосибирск     35
12  Нижний Новгород     36
13     Екатеринбург     39
14  Санкт-Петербург     42
15           Москва     54
16            Total    485

***

# выводы по анализу работодателей
В этом юните мы подробно рассматриваем работодателей. Из первого задания мы видим пять гигантов - компаний, которые выставляют больше всего вакансий (Яндекс, Ростелеком, Тинькофф, СБЕР, Газпром нефть) - на них пришлось около 7% всех вакансий. Среди них Яндекс лидирует с большим отрывом и размещает почти 2 тысячи вакансий, что осотавляет около 4% всех вакансий базы.

Можно также заметить из задания 2, что многие работодатели указывают регион "Россия" для своих компаний - но этого региона не встречается ни в одной вакансии. Это можно объяснить тем, что компания может работать по всей России, но вакансии обычно выставляются для какого-то конктретного города.

Рассмотрим количество регионов, в которых компании публикуют свои вакансии. В топах снова видим два гиганта - Яндекс и Ростелеком, которые предлагают работу даже в небольших городах.

Анализ сфер деятельности показывает, что довольно много компаний не указывают ее вовсе (более трети от общего числа), другие - указывают несколько сфер деятельности. Как пример, разработкой програмного обеспечания занимается 3553 работодателей - что составляет более 20% от числа всех работодателей, указавших хотя бы одну сферу деятельности, но очевидно, что такое высокое число работодателей из данной сферы достигается за счет того, что многие компании указывают сразу несколько сфер деятельности.

На примере самого крупного работодателя базы - компании Яндекс, посмотрим распределение вакансий в городах-миллионниках России. Москва(54) и Санкт-Петербург(42) ожидаемо в топах - но совсем немного от Санкт-Петербурга отстают и другие города-миллионники, такие как города центральной части России(Нижний Новгород(36), Воронеж(32), Краснодар(30)), а также крупнейшие по населению города за исключением Москвы и Санкт-Петербурга - Екаберинбург(39) и Новосибирск(35). Другие миллионники мало отличаются друг от друга и имеют от 20 до 30 вакансий от компании Яндекс. Всего на города-миллионники приходится 485 вакансий компании, что состовляет примерно четверть от общего числа вакансий компании Яндекс в этой базе данных.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [83]:
# текст запроса
query_6_1 = f'''select
                    count(*)
                from public.vacancies
                where (lower(name) like '%data%') or (lower(name) like '%данн%')
            '''

In [84]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [85]:
# текст запроса
query_6_2 = f'''select
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((lower(name) like '%ml%') and (lower(name) not like '%html%')))
                and ((lower(name) like '%junior%')
                or (experience = 'Нет опыта')
                or (employment = 'Стажировка'))
            '''

In [86]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [87]:
# текст запроса
query_6_3 = f'''select
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((lower(name) like '%ml%') and (lower(name) not like '%html%')))
                and ((lower(key_skills) like '%sql%')
                or (lower(key_skills) like '%postgres%'))
            '''

In [88]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [89]:
# текст запроса
query_6_4 = f'''select
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((lower(name) like '%ml%') and (lower(name) not like '%html%')))
                and ((lower(key_skills) like '%python%'))
            '''

In [90]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [91]:
# текст запроса
query_6_5 = f'''select
                    round(avg(length(key_skills) - length(replace(key_skills, CHR(9), '')) + 1), 2)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((name like '%ML%') and (lower(name) not like '%html%')))
                and (key_skills is not null)
            '''

In [92]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [93]:
# текст запроса
query_6_6 = f'''select
                    experience,
                    avg((coalesce(salary_from, salary_to) + coalesce(salary_to, salary_from))/2)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or (name like '%ML%') and (lower(name) not like '%html%'))
                and ((salary_from is not null) or (salary_to is not null))
                group by 1
            '''

In [94]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience            avg
0           Нет опыта   74642.857143
1  От 1 года до 3 лет  139674.750000
2       От 3 до 6 лет  243114.666667

***

# выводы по предметному анализу

Из всех вакансий 1771 так или иначе связаны с данными (3-4% от числа всех вакансий). При этом позиций для начинающего дата-сайентиста сильно меньше - всего 51 вакансия, и это всего 3% от числа всех вакансий по части анализа данных. Вероятно, большинство компаний хотят нанимать уже человека с опытом работы на такую должность, что может оказаться проблемой для соискателей без опыта работы в этой сфере.

В 229 вакансиях для дата сайентистов требуется навык SQL, в 357 - програмирование на Python. Среднее число ключевых навыков в вакансиях дата сайентиста - 6-7.

Из последней таблицы видно, что зарплата дата сайентиста очень зависит от опыта работы и вырастает почти в два раза за первые 1-3 года работы (140 тыс рублей при опыте 1-3 года по сравнению с 75 тыс для соискателей без опыта работы). Для соискателей с опытом в сфере от 3 до 6 лет зарплата еще выше - в среднем 243 тысячи рублей.

# Общий вывод по проекту

# дополнительные исследования данных

В первую очередь мне захотелось узнать чуть больше о вакансиях дата сайентистов. Например, в предыдущем блоке было получено количество вакансий с ключевыми навыками Python и SQL, и это число мне показалось не таким уж большим (357 и 229 соответственно). Возможно, это не самые важные навыки в данной профессии?

Для начала проверим, сколько всего вакансий есть для дата сайентистов (были использованы те же критерии, что указаны выше). Из пункта 7_2 получено, что число таких вакансий - 536. Далее я проверила число вакансий и с друкми навыками, такими как С++(46), PowerPoint(24), git(86), Data Analysis(55), numpy(41) и другие. Никакие из них по популярности не превосходят Python и SQL, но для того, чтобы сделать окончательные выводы по поводу популярности тех или иных навыков, следует сделать отдельную таблицу с названием навыка и количеством вакансий, где этот навык упоминается. Это было бы очень просто получить с помощью группировки, если бы навык указывался только один, однако в нашем случае задача оказывается более трудоемкой (можно работать в этом напрадвении для дальнейшего детального анализа вакансий).

Тем не менее в пункте 7_1 я использовала метод группировки для того, чтобы оценить популярность различных комбинаций навыков. Мы видим, что в 49 вакансиях не указано навыков вовсе, что некоторые группы навыков повторяются в описании разных вакансий (до 14 раз).

Далее я изучила подробнее, в каких регионах представлено наибольшее число вакансий для дата сайентистов. С большим отрывом идет Москва (225 вакансий) - почти половина вакансий в данной сфере размещается именно в этом городе. В Санкт-Петербурге 69 вакансий что тоже достаточно много, в других крупных городах России и ближайшего зарубежья вакансс хоть и есть, но их достаточно мало.

В завершении своего дополнительного анализа я посмотрела, какие типы занятости и графики работы наиболее популярны для нашей сферы - в пункте 7_4. С большим отрывом лидирует полная занятость - остальными результатами можно пренебречь и считать, что полная занятость - характерная черта для данной профессии. Большниство работодателей желают получить сотрудника на полный день, но есть и те, кто согласен на удаленную работу. Совсем небольшое число вакансий связано с гибким графиком - очевидно, он тоже непопулярен в данной профессии.

# итоги

В данном проекте была проанализирована база данных HeadHunter с помощью sql. Подробно исследовалось количество вакансий в различных регионах, для разных рабочих графиков, оплытов работы. Особое внимание уделено провессии дата сайентиста - получены средние заработные платы для различного опыта работы в данной сфере, исследована популярность некоторых ключевых навыков.

In [99]:
query_7_1 = f'''select
                    key_skills,
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((name like '%ML%') and (lower(name) not like '%html%')))
                group by 1
                order by 2 DESC
                limit 10
            '''

In [100]:
df = pd.read_sql_query(query_7_1, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2911634181.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_1, connection)


key_skills  count
0                                               None     49
1  Python\tMachine Learning\tData Analysis\tNumpy...     14
2  Оценка рисков\tMS PowerPoint\tСтатистический а...     13
3  Qt\tООП\tC++\tMS Visual Studio\tРазработка ком...      9
4  Python\tmachine learning\tweb security\tsklear...      9
5  Git\tООП\tAtlassian Jira\tJava\tAtlassian Conf...      7
6  Python\tSQL\tData Mining\tАналитические исслед...      6
7        Python\tLearning to Rank\tData Science\tNLP      6
8  Python\tDocker\tMachine Learning\tData Science...      5
9  TinyML\tEmbedded\tSolution Architect\tАрхитект...      4

In [97]:
query_7_2 = f'''select
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((lower(name) like '%ml%') and (lower(name) not like '%html%')))
            '''

In [98]:
df = pd.read_sql_query(query_7_2, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2993136978.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_2, connection)


count
0    536

In [103]:
query_7_3 = f'''select
                    a.name area,
                    count(*) cnt
                from public.areas a
                join public.vacancies v on a.id = v.area_id
                where ((lower(v.name) like '%data scientist%')
                or (lower(v.name) like '%data science%')
                or (lower(v.name) like '%исследователь данных%')
                or (lower(v.name) like '%machine learning%')
                or (lower(v.name) like '%машинн%обучен%')
                or ((lower(v.name) like '%ml%') and (lower(v.name) not like '%html%')))
                group by 1
                order by 2 DESC
                limit 5
            '''

In [104]:
df = pd.read_sql_query(query_7_3, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2328698441.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_3, connection)


area  cnt
0           Москва  225
1  Санкт-Петербург   69
2      Новосибирск   23
3           Алматы   22
4  Нижний Новгород   20

In [32]:
query_7_4= f'''select
                    schedule,
                    employment,
                    count(*)
                from public.vacancies
                where ((lower(name) like '%data scientist%')
                or (lower(name) like '%data science%')
                or (lower(name) like '%исследователь данных%')
                or (lower(name) like '%machine learning%')
                or (lower(name) like '%машинн%обучен%')
                or ((lower(name) like '%ml%') and (lower(name) not like '%html%')))
                group by 1, 2
                order by 3 DESC
            '''

In [33]:
df = pd.read_sql_query(query_7_4, connection)
df

C:\Users\kleop\AppData\Local\Temp\ipykernel_17056\2502122577.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_7_4, connection)


schedule           employment  count
0       Полный день     Полная занятость    356
1  Удаленная работа     Полная занятость    122
2     Гибкий график     Полная занятость     41
3     Гибкий график  Частичная занятость      5
4  Удаленная работа     Проектная работа      4
5  Удаленная работа  Частичная занятость      3
6     Гибкий график           Стажировка      2
7       Полный день           Стажировка      2
8       Полный день  Частичная занятость      1